src/systemctl/systemctl.c

```diff
@@ -6327,6 +6328,137 @@ static int log_level(int argc, char *argv[], void *userdata) {
+
+                        if (!service_name_is_valid(t))
+                                continue;
+
+                        p = TAKE_PTR(t);
+                }
+
+                if (!p)
+                        return log_error_errno(SYNTHETIC_ERRNO(ENOLINK),
+                                               "Unit %s doesn't declare BusName= or \"dbus-<destination>.service\" name.", name);
+        }
+
+        *ret_dbus_name = p;
+        return 0;
+}
+
+static int service_log_setting(int argc, char *argv[], void *userdata) {
+        sd_bus *bus;
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_free_ char *unit = NULL, *dbus_name = NULL;
+        int r;
+
+        assert(STR_IN_SET(argv[0], "service-log-level", "service-log-target"));
+        bool level = streq(argv[0], "service-log-level");
+
+        r = acquire_bus(BUS_FULL, &bus);
+        if (r < 0)
+                return r;
+
+        r = unit_name_mangle_with_suffix(argv[1], argv[0],
+                                         arg_quiet ? 0 : UNIT_NAME_MANGLE_WARN,
+                                         ".service", &unit);
+        if (r < 0)
+                return log_error_errno(r, "Failed to mangle unit name: %m");
+
+        r = service_name_to_dbus(bus, unit, &dbus_name);
+        if (r < 0)
+                return r;
+
+        if (argc == 2) {
+                _cleanup_free_ char *value = NULL;
+
+                r = sd_bus_get_property_string(
+                                bus,
+                                dbus_name,
+                                "/org/freedesktop/LogControl1",
+                                "org.freedesktop.LogControl1",
+                                level ? "LogLevel" : "LogTarget",
+                                &error,
+                                &value);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to get log %s of service %s: %s",
+                                               level ? "level" : "target",
+                                               dbus_name, bus_error_message(&error, r));
+
+                puts(value);
+
+        } else {
+                assert(argc == 3);
+
+                r = sd_bus_set_property(
+                                bus,
+                                dbus_name,
+                                "/org/freedesktop/LogControl1",
+                                "org.freedesktop.LogControl1",
+                                level ? "LogLevel" : "LogTarget",
+                                &error,
+                                "s",
+                                argv[2]);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to set log %s of service %s to %s: %s",
+                                               level ? "level" : "target",
+                                               dbus_name, argv[2], bus_error_message(&error, r));
```

Comment:
i wonder if we should do some client-side mangling or validation of at least the log level part, so that people don't willy-nilly come up with their own level definitions or so. because a generic interface that accepts non-generic parameters is not really generic. 

i.e. maybe either parse the syslog level and always pass it in number form (or only in text form). Or at least warn if it doesn't parse as number/text string. Slight preference for the former, since it means we can remove any potential confusion on whether both lowercase/uppercase should be accepted or whether the log levels in text form should be prefixed with `LOG_` or not.

Doing the same kind of mangling for the target is much harder, since the acceptable vocabulary is less clear.